### ISSUE TITLE
🎉 Asynchronous Iteration arrived at Stage 4.

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#9891]
> https://github.com/tc39/proposals/commit/a41dccc91675f41c65dee354c18c19946d6f2f45

- [x] Update Espree to support the syntax in `ecmaVersion:2018`.
    - `for await (const x of xs);` (`ForOfStatement#await`)
    - `async function* f() {}`
    - `f = async function* () {}`
    - `obj = { async* f() {} }`
    - `class A { async* f() {} }`
- [x] Update core rules to support the syntax.

**Comment by not-an-aardvark** [comment]
- [ ] `object-shorthand` fixer needs to account for async generator methods
- [ ] `semi` needs to account for `for-await-of` statements
- [ ] `keyword-spacing` needs to account for `for-await-of` statements
- [ ] `no-await-in-loop` needs to account for `for-await-of` statements (should not report statements in `for-await-of` loops)
- [ ] `no-useless-computed-key` fixer needs to account for async generator methods
- [ ] `require-await` needs to avoid reporting errors if an async function has a `for-await-of` loop

**Comment by slikts** [comment]
I really wish eslint would have a flag similar to V8's --harmony allowing the use of latest language features to avoid situations like this where the actual feature has been usable for almost a year and now is even standardized yet eslint support is lagging.

**Comment by mysticatea** [comment]
@slikts You can use `babel-eslint` parser to use experimental syntax (it's behind `--harmony` flag because it's experimental).

Now it has arrived at Stage 4, so I have implemented it ASAP: https://github.com/acornjs/acorn/pull/660.
We are awaiting the release of acorn, the upstream of our JavaScript parser.

**Comment by mysticatea** [comment]
I'm working on this.

**Comment by pkhpkh** [comment]
So, it looks like this changed the detection site for `no-await-in-loop` in `for-of` loops with included `await` statements, e.g.:

```
1 | for (x of obj) {
2 |  await foo(x);
3 | }
```

from the `await` statement to the loop. (In the example, moved from (2) to (1). This makes sense for `for-await-of` loops, but seems surprising in a `for-of` loop.

Is this intentional, or a bug?
